### ISSUE TITLE
Type inference of list literals with parameterized items does not work as expected

### ISSUE THREAD

**Issue by greglittlefield-wf** [dart-lang/sdk#34857]
Reproduced in Dart SDK Version 2.0.0 (dartpad).

I tried looking for existing issues, and found https://github.com/dart-lang/sdk/issues/32987, but I'm not sure whether this is the same.

In the example below, it would be nice if `ab_different` could be inferred as `List<A<Object>>`, just as `aa_different` is, as opposed to just `List<Object>`.

```dart
class A<T> {
  T get field => null;
}
class B<T> extends A<T> {}

main() {
  var aa_same      = [new A<int>(), new A<int>()];    // List<A<int>>
  var aa_untyped   = [new A(),      new A()];         // List<A<dynamic>>
  var aa_different = [new A<int>(), new A<String>()]; // List<A<Object>>
  
  var ab_same      = [new A<int>(), new B<int>()];    // List<A<int>>
  var ab_untyped   = [new A(),      new B()];         // List<A<dynamic>>
  var ab_different = [new A<int>(), new B<String>()]; // List<Object>
}
```

---

For context, we have some code that looks like this:
```dart
// package foo
final A<int> foo = new A<int>();
```
```dart
// package bar
final A<String> bar = new A<String>();
void doSomething(A a) => print(a.field);
[foo, bar].forEach(doSomething)
```
Changing `foo` to be a subtype of its previous value,:
```dart
// package foo
final B<int> foo = new B<int>();
```
now introduces a type error in package bar:
```dart
// The function 'doSomething' has type '(A<dynamic>) → void' that isn't of expected type '(Object) → void'. 
// This means its parameter or return type does not match what is expected.
[foo, bar].forEach(doSomething)
```

**Comment by lrhn** [comment]
This seems to be consistent behavior across front-end and analyzer.
Since both tools work this way, it's likely that it's the intended behavior.
@leafpetersen @eernstg WDYT?

**Comment by eernstg** [comment]
The least upper bound algorithm in Dart has been discussed now and then for years, and it is well-known that it does not yield a _least_ upper bound, only _some_ upper bound of its arguments. As long as we don't have union types there's no hope that we can have an algorithm that will actually provide the least upper bound:

```dart
class I {}
class J {}
class A implements I, J {}
class B implements I, J {}
```
The least upper bound of `A` and `B` could be obtained with union types as `A | B`, but we don't have union types (yet? ;-), and with intersection types we could try to use `I & J` (but that would include a class `C` that also implements `I` and `J`, so that's already "not least"), and, of course, there is no choice using the current Dart types that would be really satisfying.

So we're not going to get a perfect solution at this point.

We probably can't change anything here, because it is a breaking change which is potentially subtle and pervasive.

Anyway, we could consider using a least upper bound computation on the level of classes first, and then take the step where we compute the least upper bound of the type arguments, if any:
```dart
LUB(A<int>, B<String>) = // First step of algorithm: Find common class.
LUB(A<int>, A<String>) = // Second step of algorithm: Find common type argument.
A<Object>
```
This would work also with `class C extends B<String> {}` because we just find a common class, no matter whether no, some, or all operands of the least-upper-bound operator are parameterized types (that is: no matter whether they have actual type arguments or not). Then, when we have chosen a common class, we would proceed with the same algorithm on the type arguments, recursively.

However, even though that might seem promising, we could end up in a loop. Consider the following declarations:
```dart
class C<X extends C<X>> {}
class D extends C<D> {}
class E extends C<E> {}
```
We would then proceed as follows:
```dart
LUB(D, E) = // Step 1: Find common class.
LUB(C<D>, C<E>) = // Step 2: Find common type arguments.
LUB(C<C<D>>, C<C<E>>) = // Step 3: Find common type arguments, next level.
LUB(C<C<C<D>>>, C<C<C<E>>>) = ...
```

In summary, I'm not convinced that we have an easy path to an algorithm which is (1) not too breaking, such that we can actually use it, and (2) able to deliver "the results that anybody can immediately see we want". ;-)

**Comment by leafpetersen** [comment]
Yes, this is unfortunately working as "intended".   Least upper bound is something that I really wanted to improve from Dart 1 but couldn't find a satisfactory solution.  Maybe we should have just used the non-terminating algorithm that @eernstg describes above with an arbitrary bound on the number of steps allowed to make it terminate.  But as it is, we use an only slightly improved version of the Dart 1 algorithm, and it does in fact not find a useful LUB for the original example.

**Comment by greglittlefield-wf** [comment]
That makes sense; thanks for the detailed explanation!

The more I read about Dart's type system and type inference, the more I understand how hard of a problem it is to solve. Thanks for making it handle most cases well! ❤️
